test_rmw_implementation/test/test_graph_api.cpp

```diff
@@ -913,3 +915,151 @@ TEST_F(CLASSNAME(TestGraphAPI, RMW_IMPLEMENTATION), count_subscribers_with_bad_a
+  rmw_reset_error();
+
+  // A node from a different implementation cannot be used to query.
+  const char * implementation_identifier = node->implementation_identifier;
+  node->implementation_identifier = "not-an-rmw-implementation-identifier";
+  EXPECT_EQ(RMW_RET_INCORRECT_RMW_IMPLEMENTATION, rmw_count_clients(node, service_name, &count));
+  node->implementation_identifier = implementation_identifier;
+  rmw_reset_error();
+
+  // A null topic name is an invalid argument.
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_count_clients(node, nullptr, &count));
+  rmw_reset_error();
+
+  // An invalid topic name is an invalid argument.
+  constexpr char invalid_service_name[] = "not a valid topic name !";
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_count_clients(node, invalid_service_name, &count));
+  rmw_reset_error();
+
+  // A null count is an invalid argument.
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_count_clients(node, service_name, nullptr));
+  rmw_reset_error();
+}
+
+TEST_F(CLASSNAME(TestGraphAPI, RMW_IMPLEMENTATION), count_services_with_bad_arguments) {
+  size_t count = 0u;
+  constexpr char service_name[] = "/test_service";
+  // A null node is an invalid argument.
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_count_services(nullptr, service_name, &count));
+  rmw_reset_error();
+
+  // A node from a different implementation cannot be used to query.
+  const char * implementation_identifier = node->implementation_identifier;
+  node->implementation_identifier = "not-an-rmw-implementation-identifier";
+  EXPECT_EQ(RMW_RET_INCORRECT_RMW_IMPLEMENTATION, rmw_count_services(node, service_name, &count));
+  node->implementation_identifier = implementation_identifier;
+  rmw_reset_error();
+
+  // A null topic name is an invalid argument.
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_count_services(node, nullptr, &count));
+  rmw_reset_error();
+
+  // An invalid topic name is an invalid argument.
+  constexpr char invalid_service_name[] = "not a valid topic name !";
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_count_services(node, invalid_service_name, &count));
+  rmw_reset_error();
+
+  // A null count is an invalid argument.
+  EXPECT_EQ(RMW_RET_INVALID_ARGUMENT, rmw_count_services(node, service_name, nullptr));
+  rmw_reset_error();
+}
+
+TEST_F(CLASSNAME(TestGraphAPI, RMW_IMPLEMENTATION), count_clients_and_services) {
+  constexpr char service_name[] = "/test_service";
+  const rosidl_service_type_support_t * ts =
+    ROSIDL_GET_SRV_TYPE_SUPPORT(test_msgs, srv, BasicTypes);
+  size_t count = 0u;
+  rmw_ret_t ret;
+  rmw_service_t * srv = rmw_create_service(
+    other_node, ts, service_name, &rmw_qos_profile_services_default);
+  ASSERT_NE(nullptr, srv) << rmw_get_error_string().str;
+
+  SLEEP_AND_RETRY_UNTIL(rmw_intraprocess_discovery_delay, rmw_intraprocess_discovery_delay * 10) {
+    ret = rmw_count_services(node, service_name, &count);
+    ASSERT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
```

Comment:
This won't work, because if we fail once, that will cause everything after this to fail as well.  What we should do is remove this line from the loop, and then after the loop have an ASSERT that looks something like:

```
ASSERT_EQ(1u == count);
```

The same goes for other uses of `SLEEP_AND_RETRY_UNTIL` below.

Reply:
This tight loop checks both whether the return code is successful and whether the intended count is met. Despite this loop being used to wait for the graph update, the `rmw_count_services` should still return a successful return code even if the intended count is not reached. Therefore, it seems appropriate that a failure occurring once in this line could result in the test failing. What do you think @clalancette ?

Reply:
Oh yes, you are correct.  Thank you for pointing that out.